test/extensions/transport_sockets/alts/tsi_socket_test.cc

```diff
@@ -382,6 +383,152 @@ TEST_F(TsiSocketTest, HandshakeWithInternalError) {
+                 server_.tsi_socket_->doRead(server_.read_buffer_));
+
+  EXPECT_EQ(ClientToServerData, server_.read_buffer_.toString());
+}
+
+TEST_F(TsiSocketTest, DoReadNoData) {
+  initialize(nullptr, nullptr);
+
+  client_.write_buffer_.add(ClientToServerData);
+
+  doHandshakeAndExpectSuccess();
+
+  EXPECT_CALL(*client_.raw_socket_, doWrite(_, false));
+  expectIoResult({Network::PostIoAction::KeepOpen, 21UL, false},
+                 client_.tsi_socket_->doWrite(client_.write_buffer_, false));
+
+  EXPECT_CALL(*server_.raw_socket_, doRead(_))
+      .WillOnce(Invoke([&](Buffer::Instance& buffer) {
+        Network::IoResult result = {Network::PostIoAction::KeepOpen, client_to_server_.length(),
+                                    false};
+        buffer.move(client_to_server_);
+        return result;
+      }))
+      .WillOnce(Invoke([&](Buffer::Instance& buffer) {
+        Network::IoResult result = {Network::PostIoAction::KeepOpen, 0UL, false};
+        buffer.move(client_to_server_);
+        return result;
+      }));
+  EXPECT_CALL(server_.callbacks_, shouldDrainReadBuffer());
+  expectIoResult({Network::PostIoAction::KeepOpen, 21UL, false},
+                 server_.tsi_socket_->doRead(server_.read_buffer_));
+
+  EXPECT_EQ(ClientToServerData, server_.read_buffer_.toString());
+}
+
+TEST_F(TsiSocketTest, DoReadError) {
+  initialize(nullptr, nullptr);
+
+  client_.write_buffer_.add(ClientToServerData);
+
+  doHandshakeAndExpectSuccess();
+
+  EXPECT_CALL(*client_.raw_socket_, doWrite(_, false));
+  expectIoResult({Network::PostIoAction::KeepOpen, 21UL, false},
+                 client_.tsi_socket_->doWrite(client_.write_buffer_, false));
+
+  EXPECT_CALL(*server_.raw_socket_, doRead(_))
+      .WillOnce(Invoke([&](Buffer::Instance& buffer) {
+        Network::IoResult result = {Network::PostIoAction::KeepOpen, client_to_server_.length(),
+                                    false};
+        buffer.move(client_to_server_);
+        return result;
+      }))
+      .WillOnce(Invoke([&](Buffer::Instance& buffer) {
+        Network::IoResult result = {Network::PostIoAction::Close, 0UL, false};
+        buffer.move(client_to_server_);
+        return result;
+      }));
+  EXPECT_CALL(server_.callbacks_, shouldDrainReadBuffer());
+  expectIoResult({Network::PostIoAction::Close, 21UL, false},
+                 server_.tsi_socket_->doRead(server_.read_buffer_));
+
+  EXPECT_EQ(ClientToServerData, server_.read_buffer_.toString());
+}
+
+TEST_F(TsiSocketTest, DoReadDrainBuffer) {
+  initialize(nullptr, nullptr);
+
+  client_.write_buffer_.add(ClientToServerData);
+
+  doHandshakeAndExpectSuccess();
+
+  EXPECT_CALL(*client_.raw_socket_, doWrite(_, false));
+  expectIoResult({Network::PostIoAction::KeepOpen, 21UL, false},
+                 client_.tsi_socket_->doWrite(client_.write_buffer_, false));
+
+  EXPECT_CALL(*server_.raw_socket_, doRead(_)).WillOnce(Invoke([&](Buffer::Instance& buffer) {
```

Comment:
I would recommend adding "InSequence s;" before this line to verify that the expectations happen in the expected order.

Reply:
SG. Done.

Reply:
Done.